src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -157,122 +158,128 @@ public boolean apply(SSTableReader sstable)
+                try (Refs closedRefs = refs; CompactionController closedController = controller) {}
             }
+            Directories.removeExpectedSpaceUsedByCompaction(expectedWriteSize, CONSIDER_CONCURRENT_COMPACTIONS);
+            if (taskId != null)
+                SystemKeyspace.finishCompaction(taskId);
 
-            // log a bunch of statistics about the result and save to system table compaction_history
-            long dTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
-            long startsize = SSTableReader.getTotalBytes(transaction.originals());
-            long endsize = SSTableReader.getTotalBytes(newSStables);
-            double ratio = (double) endsize / (double) startsize;
-
-            StringBuilder newSSTableNames = new StringBuilder();
-            for (SSTableReader reader : newSStables)
-                newSSTableNames.append(reader.descriptor.baseFilename()).append(",");
+            if (collector != null && ci != null)
+                collector.finishCompaction(ci);
+        }
 
-            if (offline)
-            {
-                Refs.release(Refs.selfRefs(newSStables));
-            }
-            else
-            {
-                double mbps = dTime > 0 ? (double) endsize / (1024 * 1024) / ((double) dTime / 1000) : 0;
-                long totalSourceRows = 0;
-                String mergeSummary = updateCompactionHistory(cfs.keyspace.getName(), cfs.getColumnFamilyName(), ci, startsize, endsize);
-                logger.debug(String.format("Compacted (%s) %d sstables to [%s] to level=%d.  %,d bytes to %,d (~%d%% of original) in %,dms = %fMB/s.  %,d total partitions merged to %,d.  Partition merge counts were {%s}",
-                                           taskIdLoggerMsg, transaction.originals().size(), newSSTableNames.toString(), getLevel(), startsize, endsize, (int) (ratio * 100), dTime, mbps, totalSourceRows, totalKeysWritten, mergeSummary));
-                logger.trace(String.format("CF Total Bytes Compacted: %,d", CompactionTask.addToTotalBytesCompacted(endsize)));
-                logger.trace("Actual #keys: {}, Estimated #keys:{}, Err%: {}", totalKeysWritten, estimatedKeys, ((double) (totalKeysWritten - estimatedKeys) / totalKeysWritten));
-
-                // update the metrics
-                cfs.metric.compactionBytesWritten.inc(endsize);
-                cfs.metric.compactionsCompleted.inc();
-            }
+        try
+        {
+            ColumnFamilyStoreManager.instance.markForDeletion(cfs.metadata, transaction.logged.obsoleteDescriptors());
+        }
+        catch (Exception e)
+        {
+            logger.error("Failed to write to the write-ahead log for {}/{}. Continuing to compact now can cause resurrection. Exiting",
+                         cfs.keyspace.getName(), cfs.name, e);
+            panic();
+        }
+        refs.close();
+        controller.close();
+        if (taskId != null)
+            SystemKeyspace.finishCompaction(taskId);
```

Comment:
Do we have any tests that assert if any ancestors of a compaction are missing, the products are all left alone?

Reply:
added testRemoveUnusedSstablesDoesNotTouchProductsWhenAncestorDoesNotExist to ColumnFamilyStoreTest